_posts/2020-02-17-eda-architecture.md

```diff
@@ -0,0 +1,181 @@
+Business Services implement the customers' workflows and only focus on business rules and requirements to orchestrate and implement the value addition upon our customers' audio and data. They make use of the DP services as a library for that. There are very specific to us: you would never want to externalize your business services.
+
+Business Services persist the data they produce and are their unique trusted source of truth.
+
+Business Services build and maintain their own customer configuration from events on the bus.
+
+All services must be idempotent, that is, if they receive twice the exactly same message, they must behave identically and produce the same outputs.
+
+### Commands, Results and Events
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages emitted by business services and telling the bus what happened.
+
+A business service owns the type of Events it emits. It knows nothing about the services who will process them. It subscribes to the types of Events it needs but knows nothing about their origins.
+
+An Event type defines the meaning of the events of that type and their data schema. They must be documented.
+
+The type of an actual event message is given by its name (used as *routing key* in AMQP, see *Topology* below). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
+
+Commands are utility messages consumed by Data Processing Services.
+
+A DP Service owns the types of the commands it consumes and their results. A command is always addressed to the logical service that owns it.
+
+A Command type defines the meaning of the command, its data schema and its result data schema. It must be documented.
+
+The type of an actual command message is given by its name. It is in the form `VerbObject`. For example: `AnnotateText`, `TranscribeAudio`…
+As commands are addressed to a particular logical service, the *routing key* of a command is in the form `logical_service_name.commandname`. For example: `voxo.TranscribeAudio`.
+
+The command contains the return "address"  (`reply_to`) to which the result is to be sent and a reference set by the sender that is returned as-is (the `correlation_id`). That reference is very important for the sender: as all communications are asynchronous, the service requesting the command needs a way to reconcile the received result with the initial request it made.
+
+The `correlation_id` and `reply_to` are standard on Commands and Results, and thus, transported into the AMQP headers (in which they are already standard) instead of the message payload which is service specific.
```

Comment:
We should be consistent, if we remove AMQP earlier, it should also be removed here. Otherwise, we can keep it as is as we provided a link to AMQP protocol in previous section.